Initialize VS Code highlighting extension.

EricWay1024 opened the pull request:
This is the initial version of the VS Code highlighting extension. Here I try to structure the TextMate grammar file in the BNF way, but I've made simplifications here and there to produce a basic version for a start. Instructions on quick running, development and installation can be found in `README.md`.

EvgSkv commented on syntax/vscode/README.md:
```diff
@@ -0,0 +1,32 @@
+
+## Installation
+
+To start using your extension with Visual Studio Code copy it into the `<user home>/.vscode/extensions` folder and restart Code.
```
For users that didn't install VSCode extensions before (like myself :-) ), can you please give a command to do this?

I understand that "this extension" should be copied to the folder that you provided, but I don't know what should I copy the whole repository or yaml or json. Etc.

Having a command will save each user some time. :-)

EricWay1024 replied:
Apart from adding the `cp` command I'm also thinking about packaging and publishing the extension so that the user can install the extension more easily. The steps are described [here](https://code.visualstudio.com/api/working-with-extensions/publishing-extension#publishing-extensions), where we need a Azure DevOps organization. I was wondering if I should create the organization myself or you can help me with this.

EvgSkv commented:
This is great! I've left a few minor comments, please take a look.

I tried it in my VSCode and highlighting already looks very reasonable. I have some  comments on that, but I am not sure if code as-is should already be coloring everything, or it's a draft so far.

E.g. in expression
```
P := F(A:B)
```
The predicate `P` is colored gray, but I think it should be same  color as other predicates.

Let me know if you need more details/screenshots on this, or this is something you wanted to polish later.

EricWay1024 commented on syntax/vscode/src/logica.tmLanguage.yaml:
```diff
@@ -0,0 +1,403 @@
+        match: "@"
+      - name: keyword.operator.spread.logica
+        match: "\\.\\."
+      - name: keyword.operator.assignment.logica
```
I actually pick from the standard names which determines the coloring. The standard names include
- `keyword.operator.logical`
- `keyword.operator.assignment`
- `keyword.operator.spread`
- `keyword.operator.arithmetic`
- `keyword.operator.comparison`
- `keyword.operator.bitwise`

In my opinion semantically `:-` and `:=` belong to `keyword.operator.assignment` :-)

EvgSkv replied:
Yes, if it's a standard names, then I agree with you. Thanks for clarifying!

EricWay1024 commented:
Thank you for your review! I forgot to make it clear that the code as-is is what I need to polish later on, sorry for the confusion caused. Things need to improve include:
- Implementing the syntax of `combine`, `assign_combination` and `functor_application`
- Changing places where the syntax has been simplified, i.e. where I only included patterns involved but didn't specify the structure how the patterns are combined
- Refinements on sub-patterns of `expression` and `proposition` to make the recursive patterns work correctly
- Fixing potential bugs within the code

I commented out several places where I used simplified/temporary methods to make the highlighting work, but they need further changes and testings.

So this is actually only a starting version of the extension :-) I'm submitting it now because I'm getting busy in May with my exams and I may not have enough time for this. I'll start to spend more time on fixing these problems in June.

EvgSkv commented:
Thanks for replies! This is indeed a great start!
As you suggested, I'll create the Azure org in nearest days and will let you know.

Do you want to add the install command to the current pull request?

Thanks again for doing it so quickly!

EvgSkv commented:
@EricWay1024 , I have created an Azure DevOps, feel free to email me at logica@evgskv.dev, if you'd like me to add you to the org.
But if you prefer I can follow your instructions for adding the extension.

EricWay1024 commented:
Would you like to add me to the org please? My email is `smyyw8@nottingham.edu.cn`. (Do I have to send you an email for this?) I think there are several steps to take in the packaging and publishing process which I'm not yet familiar with, so I'll have a try before including them in the instruction. (Sorry for this late reply)

EvgSkv commented:
> Would you like to add me to the org please? My email is `smyyw8@nottingham.edu.cn`. (Do I have to send you an email for this?) I think there are several steps to take in the packaging and publishing process which I'm not yet familiar with, so I'll have a try before including them in the instruction. (Sorry for this late reply)

@EricWay1024 , I've added you to the org! Please let me know if you got the invitation and if some kind of other rights in the org are needed.
Thank you for working on this!

EricWay1024 commented:
Invitation received! I've joined the org. I'll let you know if I need further help.
Thanks a lot!

EricWay1024 commented:
I think it is also necessary for you to create a publisher in [here]( https://marketplace.visualstudio.com/manage/publishers/) and add my email `smyyw8@nottingham.edu.cn` to members of the publisher. Thanks!

EvgSkv commented:
@EricWay1024 I created a publisher, but for some reason I am unable to add any new accounts there, getting "Invalid domain" error. I wrote to support about it.
Meanwhile, it looks like VSIX file is needed to upload an extension.
If you could tell me how to obtain it, or send it to me, I'll be happy to try it before we figure out how to add your account to the publisher.
Thank you!

EricWay1024 commented:
@EvgSkv Looks like a problem described [here](https://stackoverflow.com/questions/56032912/vs-marketplace-add-member-displayes-invalid-domain-error). As is mentioned in the top answer, my VSID is `57a2c8a0-1735-6012-b747-593efeb4da18`, which you may need to inform Microsoft support of.

As of the VSIX file, it appears that once I've been added to the publisher member, I will be authorized to package and upload extensions from my command line, so I don't think it's a problem here. Actually the whole thing is like this: to package (generate) a VSIX file, I need to include a "publisher" field in the `package.json` config file, and `vsce` program will check if the account I log in with (via a personal access token) belongs to the publisher. If yes, it will include the publisher name as part of the signatures of the package file. Therefore it appears that unless my account belongs to a publisher, I'm unable to package a VSIX file. 

I'll take care of the packaging and uploading if you could resolve the adding member problem. Thank you!

EvgSkv commented:
@EricWay1024 Thank you! I've written to their support, hopefully they'll add your account soon.

EricWay1024 commented:
I've been successfully added to the publisher! Now I've published the first version. 

I've also added instructions on packaging and publishing. I moved development instructions to `DEV.md` because `README.md` is going to be displayed on the Marketplace as an extension introduction/documentation etc. for users.

EvgSkv commented:
Excellent! I've installed it for myself and it works! Thank you very much!

Is the pull request ready to be merged, or did you want to add/edit anything else?

EricWay1024 commented:
I think I'll leave tweaks for another pull request so you can merge it now :) thanks a lot

EvgSkv commented:
@EricWay1024  Sounds good! I will merge the pull request now. Thank you very much for your contribution!
